Add vector visualizer

HansOlsson opened the pull request:
This is an attempt related to #1891.
However, there are a number of issues:

- It should be better coordinated with the Arrow-visualizer - or replace it.
- It might be refactored in a better way as a new object in ModelicaServices.

tobolar commented:
I think #1891 talks about shape type whereby you add a new visualizer similar to `Arrow`. What is the benefit?

beutlich commented on Modelica/Mechanics/MultiBody/Types.mo:
```diff
@@ -117,7 +117,7 @@ are shown for a cylinder:
 </html>"));
   type ShapeType = Modelica.Icons.TypeString
-    "Type of shape (box, sphere, cylinder, pipecylinder, cone, pipe, beam, gearwheel, spring, <external shape>)"
+    "Type of shape (box, sphere, cylinder, pipecylinder, cone, pipe, beam, gearwheel, spring, vector, <external shape>)"
```
Can we have it between gearwheel (108) and spring (111)?

HansOlsson replied:
Done.

HansOlsson replied:
And now removed as part of redesign. So complete resolved.

HansOlsson commented:
> I think #1891 talks about shape type whereby you add a new visualizer similar to Arrow. 

Yes, that is exactly what is intended here.

> What is the benefit?

The main benefit of having a different shape type (and just one shape - instead of trying to build an arrow from different parts) is that it is easy to adjust the force-(etc)-visualization after a simulation, in particular this would allow the force-visualization to be active in more cases, but as default scaled to zero.

Currently the force-visualizations in MSL are mostly de-activated. I believe one reason is that it gives so little benefit compared to the effort of tuning the N-to-m scale - and thus the simplest choice is to de-activate it.

tobolar commented:
Ok I understand the benefit of the new shape type now. 

I think it is wise to replace the "old" `Arrow` and `DoubleArrow` with `Vector` in the future major version, with `Vector` being renamed to `Arrow`.

I also think there should be another `VectorQuantity` called "axis" and maybe something like "others". The "axis" could be used e.g. for visualization of coordinate axes in `World` and `FixedFrame`.

HansOlsson commented:
> The selection of shapeType="vector" would be possible in Modelica.Mechanics.MultiBody.Visualizers.Advanced.Shape as well. But the handling of parameters length, width and height would be confusing for user for that shape type.

One possibility would be to remove it from that list as follows:

Originally I considered having it as a separate entity with a separate tool-specific implementation in ModelicaServices - and even if it generates a bit more work we could do that. That Dymola implements it using a specific shapeType shouldn't matter in that case.

tobolar commented:
> Originally I considered having it as a separate entity ...

and 

> easy to adjust the force-(etc)-visualization after a simulation

... I guess the idea is good. But the usage must be straightforward to users. To my opinion this is not granted at the moment.

HansOlsson commented:
I realized that I should have returned to this earlier.

After some additional thought I believe it is best to have this as a completely separate object in ModelicaServices; corresponding to "VisualVector.mo" in Dymola (that allows re-adjusting the scaling after a simulation etc). 

I could change to that quickly (maybe as a new PR since several changes would be reverted and it might be simpler to start from scratch).

However, that seems like a new addition and anyone else implementing ModelicaServices would need to implement something for this.  

I assume that is a bit late in the release process for this release, or?

beutlich commented:
> 
> 
> I realized that I should have returned to this earlier.
> 
> After some additional thought I believe it is best to have this as a completely separate object in ModelicaServices; corresponding to "VisualVector.mo" in Dymola (that allows re-adjusting the scaling after a simulation etc).
> 
> I could change to that quickly (maybe as a new PR since several changes would be reverted and it might be simpler to start from scratch).
> 
> However, that seems like a new addition and anyone else implementing ModelicaServices would need to implement something for this.
> 
> I assume that is a bit late in the release process for this release, or?

@HansOlsson This PR is open and resolved far too long. Do you think yu can revise your proposal according to your latest ideas?

HansOlsson commented:
> I realized that I should have returned to this earlier.
> After some additional thought I believe it is best to have this as a completely separate object in ModelicaServices; corresponding to "VisualVector.mo" in Dymola (that allows re-adjusting the scaling after a simulation etc).
> I could change to that quickly (maybe as a new PR since several changes would be reverted and it might be simpler to start from scratch).
> However, that seems like a new addition and anyone else implementing ModelicaServices would need to implement something for this.
> I assume that is a bit late in the release process for this release, or?
> 
> @HansOlsson This PR is open and resolved far too long. Do you think yu can revise your proposal according to your latest ideas?

Yes should be straightforward, but maybe it will be after the summer vacation. I didn't act previously since it didn't seem relevant for the 3.2.3 release.

HansOlsson commented:
Seems I messed up the merging of other changes. Will redo in some way.

HansOlsson commented:
> Seems I messed up the merging of other changes. Will redo in some way.
Or not?

If I look here way too many files are changed, but if I look at my branch "HansOlsson/AddVector" and Compare it only shows 6 files changed, and the changes seem correct.

I guess a squash-merge (or rebase before merging) will work in this case.

What I did was:
- Base it on 4.0.0
- Updated the documentation to both explain the difference to Arrow, and be more consistent with other documentation. 
- Not modified the existing Arrow and DoubleArrow.

Note: Tool-specific ModelicaServices will need to be updated, but I believe we have time for that before 4.0.0 is released.

HansOlsson commented:
> For me it is rather confusing to have an additional vectorAnimation flag. I'd prefer to have it all animated with merely one animation choice, the one we already have.

Is this still relevant? If so can easily do that

tobolar commented:
> ...I can force-push to your branch HansOlsson/AddVector here.

@HansOlsson I will wait with my review for your replay.

HansOlsson commented:
> If you agree on the changes I can force-push to your branch HansOlsson/AddVector here.

Yes, please do.

beutlich commented:
> 
> 
> > If you agree on the changes I can force-push to your branch HansOlsson/AddVector here.
> 
> Yes, please do.

Done.

HansOlsson commented:
@beutlich Thanks for the squash.
@tobolar should I remove the vectorAnimation-parameter?

HansOlsson commented:
Hi @tobolar I just wanted to verify that you still want to remove the vectorAnimation-parameter?

beutlich commented:
@tobolar Can you please review! Thanks.

tobolar commented:
Concerning `Sensors.CutForce` and similar models, is it reasonable to enable visualization by `Arrow`or `Vector`? I suppose Vector only shall be used, shouldn't it? Then, `vectorAnimation` is really a surplus.

The reference frame of Vector is always being at its tail, right? There shall be an option to place Vector's head at the reference frame instead. This could be helpful e.g. for forces.

HansOlsson commented:
> Concerning Sensors.CutForce and similar models, is it reasonable to enable visualization by Arrowor Vector? I suppose Vector only shall be used, shouldn't it? Then, vectorAnimation is really a surplus.

That is the tricky issue, which depends on how much it is used. Assuming it isn't used much and it is ok with you @tobolar  then I can remove the old Vector-animation.

And if we remove the existing Vector-animation, then we should also remove a number of a parameters (as they are part of the post-processing and can be changed uniformly), which will simplify the model - but a bit annoying if people used had set them.

To see if it matters: in MSL there is only one use of these Cut* models, and that one has animation=false. Loading some commerical libraries I found one more - also with animation=false.

> The reference frame of Vector is always being at its tail, right? There shall be an option to place Vector's head at the reference frame instead. This could be helpful e.g. for forces.

That's the "pushing" parameter. Maybe it should have a better name, and possibly be exposed for the Cut* models.

tobolar commented:
> we should also remove a number of parameters

I guess only `forceDiameter` and `N_to_m` are affected. And probably only the latter will be crucial. Maybe it could be nice to have some strategy how to convert/retain the functionality of this parameter. Or is this tool specific?

The `VectorQuantity` shall be extended to position vector as well. This possibility shall not be excluded since it enhances the usability. If agreed, the documentation shall be extended accordingly.

The icon of the `Vector` could be a spatial arrow, designed similarly to current `VoluminousWheel` or `Torus`.

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/modelica/ModelicaStandardLibrary?pullRequest=2501) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/modelica/ModelicaStandardLibrary?pullRequest=2501) before we can accept your contribution.<br/>**0** out of **3** committers have signed the CLA.<br/><br/>:x: HansOlsson<br/>:x: beutlich<br/>:x: tobolar<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/modelica/ModelicaStandardLibrary?pullRequest=2501) it.</sub>

HansOlsson commented:
Finally returning to this.

> we should also remove a number of parameters
> 
> I guess only forceDiameter and N_to_m are affected. And probably only the latter will be crucial. Maybe it could be nice to have some strategy how to convert/retain the functionality of this parameter. Or is this tool specific?

I believe it would be tool specific, and in particular I think that changing all of the scalings at the same time is important - and thus handling the individual component-settings isn't straightforward.

> The VectorQuantity shall be extended to position vector as well. This possibility shall not be excluded since it enhances the usability. If agreed, the documentation shall be extended accordingly.

Will try to do that. 

Do you have any good names for that as it isn't really a "position", but a position-difference at a specific position. 

The corresponding sensor is called "relative position"; I will use that if there isn't anything better.
(It isn't "distance" since that is normally a scalar, and the other possible names such as "translation", "displacement" don't seem right.)

> The icon of the Vector could be a spatial arrow, designed similarly to current VoluminousWheel or Torus.

Ok.

HansOlsson commented:
It is still WIP for the proposed changes:

- [ ] Ensure that it is used inside MSL.
- [ ] Provide conversions and remove old variants. (I will turn this into a separate commit in case we want to reconsider.)

The new icon isn't the best 3d-look, but to me it works well in the diagrams.

HansOlsson commented:
This should now address the reported issues, and be a working idea (and pass the CI-tests).
It might be that some additional simplifications are possible.

It unfortunately requires an updated Dymola-specific ModelicaServices and even updated Dymola  to test in a tool.

dietmarw commented:
@tobolar Is it now fine for you?

tobolar commented:
> Is it now fine for you?

Unfortunately not yet - see my comments below.
1. What with `MultiBody.Visualizers.Advanced.DoubleArrow`? It was used in all torque elements (`Forces.WorldTorque`, `Forces.Torque`, etc.) and is still so documented in those elements. A possibility is to enable a double arrow option directly in `Vector`element.

2. Variable `diameter` in `Modelica.Mechanics.MultiBody.Visualizers.Advanced.Arrow` is no more used. It shall be deleted with corresponding conversion.

3. Rename input `r_value` in `Modelica.Utilities.Internal.PartialModelicaServices.Animation.PartialVector` to: 
a. coordinates or
b. vectorCoordinates or
c. vector.
(With my descending preference.) See also e.g. https://www.math24.net/vector-coordinates/

4. Rename accodingly `r_head` in `Modelica.Mechanics.MultiBody.Visualizers.Advanced.Arrow`.

beutlich commented:
@HansOlsson @tobolar Do you think this PR will be ready to merge by next Wednesday (29th of Jan.)?

tobolar commented:
@beutlich I guess the most essential is point 1: 
> What with MultiBody.Visualizers.Advanced.DoubleArrow?

And the work to do depends on @HansOlsson availability.

HansOlsson commented:
> @beutlich I guess the most essential is point 1:
> 
> > What with MultiBody.Visualizers.Advanced.DoubleArrow?

The simplest would be to add "doubleArrow" as a flag for the new vector (that would be fairly simple), and we then have a choice between:
* Remove DoubleArrow model
* Base DoubleArrow on the new vector (i.e. just setting a flag for a component).
 Which one is preferred?

> And the work to do depends on @HansOlsson availability.

tobolar commented:
> - Remove DoubleArrow model
> - Base DoubleArrow on the new vector

Whatever decission will be met, this should be done consistently to `Arrow`.

HansOlsson commented:
I believe it should be ready now, as I have:

- Based DoubleArrow on new vector-animation.
- Removed diameter consistently (previously it was only partially done in conversion scripts). I have deliberately not removed parameters from the World-object, as that might be seen as global scene settings.
- Renamed r_value in the new Vector to coordinates.

I have not renamed r_head in Arrow and DoubleArrow; as it wasn't strictly needed - and changing old code just before the dead-line is always risky.

If we think it odd that the classes are so similar we could make them more different by making Vector even more low-level by removing 'R' (and perform the transformation in Arrow).

tobolar commented:
> If we think it odd that the classes are so similar we could make them more different by making Vector even more low-level by removing 'R' (and perform the transformation in Arrow).

IMO the best is to keep "r" and "R" together in one element. So no change is needed here.

beutlich commented on Modelica/Utilities/Internal.mo:
```diff
@@ -50,6 +50,36 @@ This model is documented at
+        "Material property describing the reflecting of ambient light (= 0 means, that light is completely absorbed)"
+        annotation(Dialog);
+      input Boolean pushing=false "= true, if the vector is pointing towards the origin" annotation(Dialog);
+      input Boolean doubleArrow=false "= true, if the arrow has two heads after each other (pointing in the same direction)" annotation(Dialog);
```
I do not like the names `pushing` and `doubleArrow` that much.

1. Should we rename `pushing` to something like `inwardArrow` or `arrowDirection` or ...?
1. `doubleArrow` could also mean :left_right_arrow:. Should we rename it to `twoHeadedArrow` to avoid disambiguation?

tobolar replied:
> 1. pushing is associated with force arrows and not general enough. Should we rename pushing to something like inwardArrow or arrowDirection or ...?

Actually, the vector is given by its coordinates having the origin in the "vector frame". So generally, the vector is always pointing out of the vector's frame, i.e. the vector's tail is identical to origin and vector's head it given by its coordinates. What is changed by "pushing" is the head/tail assignment. So that input's name could somehow addres this. But I have no good idea.
In `Modelica.Mechanics.MultiBody.Sensors.CutForce` there is used `positiveSign` but it is not suitable here. 


> 2. doubleArrow could also mean ↔️. Should we rename it to twoHeadedArrow to avoid disambiguation?

or "doubleArrowHead" or simply "doubleHead"...

beutlich replied:
Hm, is `pushing` still used after all or is it a left-over from development?

HansOlsson replied:
It is still used.

tobolar replied:
What about "tailAtOrigin" or "headByCoordinates" ...

beutlich replied:
Why not `headAtOrigin`?

HansOlsson replied:
I think "headAtOrigin" would work, and similarly twoHeadedArrow.

tobolar replied:
@HansOlsson Shall I change it?

tobolar replied:
> Why not headAtOrigin?

Due to original setting of `pushing=false` - so to my understanding the tail was at origin. But no problem about it.

HansOlsson replied:
But "headAtOrigin" and "pushing" should mean the same, right?
I.e. you are pushing at the origin when the head is there.

tobolar replied:
Actually yes. You're right.

HansOlsson commented:
Please go ahead!

Best Regards,

Hans OLSSON

CATIA Systems/Dymola R&D Development Senior Manager





Office: +46 4 6270 6710
hans.olsson@3ds.com<mailto:Hans.OLSSON@3ds.com>

[3DS Logo]

3DS.COM/CATIA<http://www.3ds.com/CATIA>


Dassault Systemes AB | IDEON Gateway Scheelevägen 27 | | Sweden


From: tobolar <notifications@github.com>
Sent: Tuesday, January 28, 2020 16:21
To: modelica/ModelicaStandardLibrary <ModelicaStandardLibrary@noreply.github.com>
Cc: OLSSON Hans <Hans.OLSSON@3ds.com>; Mention <mention@noreply.github.com>
Subject: Re: [modelica/ModelicaStandardLibrary] AddVectorVisualizer (#2501)


@tobolar commented on this pull request.

________________________________

In Modelica/Utilities/Internal.mo<https://github.com/modelica/ModelicaStandardLibrary/pull/2501#discussion_r371867807>:

> +      input Boolean pushing=false "= true, if the vector is pointing towards the origin" annotation(Dialog);

+      input Boolean doubleArrow=false "= true, if the arrow has two heads after each other (pointing in the same direction)" annotation(Dialog);

@HansOlsson<https://github.com/HansOlsson> Shall I change it?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/modelica/ModelicaStandardLibrary/pull/2501?email_source=notifications&email_token=ADE25FXAYOHXS3O3BFGF2X3RABENFA5CNFSM4E2HWLT2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCTKFKNA#discussion_r371867807>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ADE25FR3BE2K4TXWJ5YZHG3RABENFANCNFSM4E2HWLTQ>.

This email and any attachments are intended solely for the use of the individual or entity to whom it is addressed and may be confidential and/or privileged.

If you are not one of the named recipients or have received this email in error,

(i) you should not read, disclose, or copy it,

(ii) please notify sender of your receipt by reply email and delete this email and all attachments,

(iii) Dassault Systèmes does not accept or assume any liability or responsibility for any use of or reliance on this email.


Please be informed that your personal data are processed according to our data privacy policy as described on our website. Should you have any questions related to personal data protection, please contact 3DS Data Protection Officer at 3DS.compliance-privacy@3ds.com<mailto:3DS.compliance-privacy@3ds.com>


For other languages, go to https://www.3ds.com/terms/email-disclaimer

beutlich commented:
@tobolar @HansOlsson I added three more commits. Please cross-check before the planned merge. Thank you.

tobolar commented:
@beutlich Generally, it is a better option if the tool arranges all inputs, parameters, etc. automatically for the documentation. But I can live with your changes.

HansOlsson commented:
> 
> 
> @tobolar @HansOlsson I added three more commits. Please cross-check before the planned merge. Thank you.

Seems ok.

HansOlsson commented:
> 
> 
> > @tobolar @HansOlsson I added three more commits. Please cross-check before the planned merge. Thank you.
> 
> Seems ok.

Wait. Something is broken.

tobolar commented:
> Wait. Something is broken.

I will add minor changes on documentation as well.

tobolar commented:
Fine for me now.

HansOlsson commented:
> 
> 
> Fine for me now.

And now fine for me as well.
